### PR TITLE
Add logging that matches SignalProtocolKit (the ObjC library)

### PR DESCRIPTION
**Pull request by jrose-signal**

### CODE REVIEW

File: rust/protocol/src/session_cipher.rs [hunk 1]
```diff
@@ -96,6 +104,10 @@ pub async fn message_encrypt(
         )
         .await?
     {
+        log::warn!(
+            "Previously known identity key for while encrypting for recipient: {}",
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:108]: I found this phrasing really hard to parse. "for while encrypting"?

How about "Identity key {} is not trusted for remote address {}"?

File: rust/protocol/src/session_cipher.rs [hunk 2]
```diff
@@ -228,6 +240,10 @@ pub async fn message_decrypt_signal<R: Rng + CryptoRng>(
...
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:244]: OK I see these are directly from AxolotlKit but I still think they are not even English. :)

File: rust/protocol/src/session_cipher.rs [hunk 3]
```diff
@@ -296,21 +312,53 @@ fn decrypt_message_with_record<R: Rng + CryptoRng>(
     ciphertext: &SignalMessage,
     csprng: &mut R,
 ) -> Result<Vec<u8>> {
+    let log_decryption_failure = |state: &SessionState, error: &SignalProtocolError| {
+        log::error!(
+            "Failed to decrypt whisper message with ratchet key: {} and counter: {}. \
+             Session loaded for {}. Local session has base key: {} and counter: {}. {}",
+            hex::encode(ciphertext.sender_ratchet_key().public_key_bytes().unwrap()),
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:319]: Is this `unwrap` really safe?

Might be worth adding a util that hex encodes an `Optional<AsRef<[u8]>>` that returns something like "{missing}" when `None`?

File: rust/protocol/src/session_cipher.rs [hunk 4]
```diff
@@ -296,21 +312,53 @@ fn decrypt_message_with_record<R: Rng + CryptoRng>(
...
         match result {
             Ok(ptext) => {
+                log::debug!(
+                    "successfully decrypted with current session state (base key {})",
+                    hex::encode(
+                        current_state
+                            .sender_ratchet_key()
+                            .unwrap()
+                            .public_key_bytes()
+                            .unwrap()
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:351]: Again concerned about the unwraps though these are probably safe since if we were able to decrypt the key is probably there...

File: rust/protocol/src/state/session.rs [hunk 5]
```diff
@@ -210,6 +210,10 @@ impl SessionState {
         self.session.receiver_chains.push(chain);
 
         if self.session.receiver_chains.len() > consts::MAX_RECEIVER_CHAINS {
+            log::info!(
+                "Trimming excessive receiver_chain count: {}",
```
Comment (jack-signal) [inline rust/protocol/src/state/session.rs:214]: Worth logging a session id here?

File: rust/protocol/src/session_cipher.rs [hunk 6]
```diff
@@ -342,6 +401,20 @@ fn decrypt_message_with_record<R: Rng + CryptoRng>(
...
+                "No valid session for recipient: {}, current session base key {}, number of previous states: {}",
+                remote_address,
+                hex::encode(current_state.sender_ratchet_key().unwrap().public_key_bytes().unwrap()),
+                record.previous_session_states().unwrap().count(),
+            );
+        } else {
+            log::error!(
+                "No valid session for recipient: {}, (no current session state), number of previous states: {}",
+                remote_address,
+                record.previous_session_states().unwrap().count(),
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:415]: Worth adding to `SessionState` a

`pub(crate) fn previous_session_state_count() -> usize`?

File: rust/protocol/src/session_cipher.rs (hunk 4 above)
Comment (jack-signal) [reply rust/protocol/src/session_cipher.rs:351]: This specific value is logged regularly - maybe add a `fn` to `SessionState` to handle it?

File: rust/protocol/src/session_cipher.rs [hunk 7]
```diff
@@ -322,17 +370,28 @@ fn decrypt_message_with_record<R: Rng + CryptoRng>(
...
         match result {
             Ok(ptext) => {
+                log::info!(
+                    "successfully decrypted with PREVIOUS session state (base key {})",
+                    hex::encode(
+                        previous
+                            .sender_ratchet_key()
+                            .unwrap()
+                            .public_key_bytes()
+                            .unwrap()
```
Comment (jack-signal) [inline rust/protocol/src/session_cipher.rs:384]: Could (imo should) be a `fn` on `SessionState` to return this as above

**Review by jack-signal — COMMENTED** [review]
Same few items are pulled from SessionState worth adding a fn to encapulsate that and also any use of `unwrap` makes me nervous.

Especially because I'm not totally sure how much information actually makes it way to the app log if we panic... if all we see is "unwrap failed" we are going to be sad and it will be a pita to debug. At the least we should use `expect` everywhere with a reason of why we believe the value will never be Err/None.

File: rust/protocol/src/session_cipher.rs (hunk 6 above)
Comment (jrose-signal) [reply rust/protocol/src/session_cipher.rs:415]: Honestly either `previous_session_states` shouldn't be failable or it should be handled here.

**Review by jack-signal — APPROVED** [review]
